### ISSUE TITLE
Violation of implicit ordering guarantees of query commands inside ValidationStateTracker can cause false positive

### ISSUE THREAD

**Issue by psalz** [KhronosGroup/Vulkan-ValidationLayers#1458]
I've stumbled upon a case where issuing query commands after a call to `vkCmdResetQueryPool` still produces warnings about queries not being reset, when combined with queue synchronization operations such as `vkWaitForFences` or `vkQueueWaitIdle`.

Essentially what happens is that while the state tracker correctly updates the global query state maps upon command buffer submission (thus modeling the implicit ordering guarantees of query commands), queue synchronization can cause newer query states to be overwritten with older ones. I think it's easiest explained by first looking at a code sample (I've repurposed the `VkPositiveLayerTest.ResetQueryPoolFromDifferentCB` test to reproduce this issue, so you should be able to just copy/paste this into the test suite -- but never mind the name):

```cpp
TEST_F(VkPositiveLayerTest, ResetQueryPoolFromDifferentCB) {
    TEST_DESCRIPTION("Reset a query on one CB and use it in another.");

    ASSERT_NO_FATAL_FAILURE(Init());

    m_errorMonitor->ExpectSuccess();

    VkFence ts_fence;
    VkFenceCreateInfo fence_info{};
    fence_info.sType = VK_STRUCTURE_TYPE_FENCE_CREATE_INFO;
    fence_info.flags = VK_FENCE_CREATE_SIGNALED_BIT;
    vk::CreateFence(m_device->device(), &fence_info, nullptr, &ts_fence);

    VkQueryPool query_pool;
    VkQueryPoolCreateInfo query_pool_create_info{};
    query_pool_create_info.sType = VK_STRUCTURE_TYPE_QUERY_POOL_CREATE_INFO;
    query_pool_create_info.queryType = VK_QUERY_TYPE_TIMESTAMP;
    query_pool_create_info.queryCount = 1;
    vk::CreateQueryPool(m_device->device(), &query_pool_create_info, nullptr, &query_pool);

    VkCommandBuffer command_buffer[2];
    VkCommandBufferAllocateInfo command_buffer_allocate_info{};
    command_buffer_allocate_info.sType = VK_STRUCTURE_TYPE_COMMAND_BUFFER_ALLOCATE_INFO;
    command_buffer_allocate_info.commandPool = m_commandPool->handle();
    command_buffer_allocate_info.commandBufferCount = 2;
    command_buffer_allocate_info.level = VK_COMMAND_BUFFER_LEVEL_PRIMARY;
    vk::AllocateCommandBuffers(m_device->device(), &command_buffer_allocate_info, command_buffer);

    {
        VkCommandBufferBeginInfo begin_info{};
        begin_info.sType = VK_STRUCTURE_TYPE_COMMAND_BUFFER_BEGIN_INFO;
        begin_info.flags = VK_COMMAND_BUFFER_USAGE_SIMULTANEOUS_USE_BIT;

        vk::BeginCommandBuffer(command_buffer[0], &begin_info);
        vk::CmdResetQueryPool(command_buffer[0], query_pool, 0, 1);
        vk::EndCommandBuffer(command_buffer[0]);

        vk::BeginCommandBuffer(command_buffer[1], &begin_info);
        vk::CmdWriteTimestamp(command_buffer[1], VK_PIPELINE_STAGE_BOTTOM_OF_PIPE_BIT, query_pool, 0);
        vk::EndCommandBuffer(command_buffer[1]);
    }

    VkSubmitInfo submit_info{};
    submit_info.sType = VK_STRUCTURE_TYPE_SUBMIT_INFO;
    submit_info.commandBufferCount = 1;
    submit_info.pCommandBuffers = nullptr;
    submit_info.signalSemaphoreCount = 0;
    submit_info.pSignalSemaphores = nullptr;

    // Begin by resetting the query pool.
    {
        submit_info.pCommandBuffers = &command_buffer[0];
        // This causes the global query state map to be updated from QUERYSTATE_UNKNOWN to QUERYSTATE_RESET.
        vk::QueueSubmit(m_device->m_queue, 1, &submit_info, VK_NULL_HANDLE);
    }

    // Write a timestamp, and add a fence to be signalled.
    {
        submit_info.pCommandBuffers = &command_buffer[1];
        vk::ResetFences(m_device->device(), 1, &ts_fence);
        // This verifies that the local/global query state is QUERYSTATE_RESET (which it is) and updates the state to QUERYSTATE_ENDED.
        vk::QueueSubmit(m_device->m_queue, 1, &submit_info, ts_fence);
    }

    // Reset query pool again.
    {
        submit_info.pCommandBuffers = &command_buffer[0];
        // Update global query state map to QUERYSTATE_RESET again.
        vk::QueueSubmit(m_device->m_queue, 1, &submit_info, VK_NULL_HANDLE);
    }

    // Finally, write a second timestamp, but before that, wait for the fence.
    {
        submit_info.pCommandBuffers = &command_buffer[1];
        // Here is the problem:
        // - Inside RetireWorkOnQueue, all command buffers up until the fence are visited again.
        // - Both previous commands (reset followed by timestamp) cause updates in the *local* query state map.
        // - However, the timestamp also causes an update in the *global* query state map, overwriting the state
        //   change from the query pool reset queue submission that happened after the fence.
        vk::WaitForFences(m_device->device(), 1, &ts_fence, true, std::numeric_limits<uint64_t>::max());
        // This check now fails as the global query state contains QUERYSTATE_AVAILABLE.
        vk::QueueSubmit(m_device->m_queue, 1, &submit_info, VK_NULL_HANDLE);
    }

    vk::QueueWaitIdle(m_device->m_queue);

    vk::FreeCommandBuffers(m_device->device(), m_commandPool->handle(), 2, command_buffer);
    vk::DestroyQueryPool(m_device->device(), query_pool, nullptr);
    vk::DestroyFence(m_device->device(), ts_fence, nullptr);

    m_errorMonitor->VerifyNotFound();
}
```

So what happens is:
1) The initial `vkCmdResetQueryPool` submission causes the global query state map to be initialized from `QUERYSTATE_UNKNOWN` to `QUERYSTATE_RESET` inside `ValidationStateTracker::PostCallRecordQueueSubmit` [here](https://github.com/KhronosGroup/Vulkan-ValidationLayers/blob/2be3fe0f84713f6ce914886279f10f3ba16aaaff/layers/state_tracker.cpp#L1439).
2) The first call to `vkCmdWriteTimestamp` verifies that the state is `QUERYSTATE_RESET` ([here](https://github.com/KhronosGroup/Vulkan-ValidationLayers/blob/2be3fe0f84713f6ce914886279f10f3ba16aaaff/layers/core_validation.cpp#L6896)), which it is. It then update the state to `QUERYSTATE_ENDED`.
3) Then we reset again.
4) Before the second call to `vkCmdWriteTimestamp`, we wait for the fence submitted with the first call. This causes the first reset and timestamp commands to be visited again inside `ValidationStateTracker::RetireWorkOnQueue`. Now here is the problem: If any of the query update functions produces `QUERYSTATE_ENDED` as the new state (which the first timestamp does), this will be written into the global query state map [here](https://github.com/KhronosGroup/Vulkan-ValidationLayers/blob/2be3fe0f84713f6ce914886279f10f3ba16aaaff/layers/state_tracker.cpp#L1317). This effectively overwrites the reset in step 3 and causes the subsequent call to `vkCmdWriteTimestamp` to fail the state check.

---

This might be the cause of #1179, but as I'm not sure and I'm pretty confident I have pinned down the exact problem in this case, I decided to open a new issue.

**Comment by DiligentGraphics** [comment]
I believe we experience the same problem: we create a query pool and then immediately reset it with `vkCmdResetQueryPool`. However, sometimes we get the "query must be reset before it is use" error.

**Comment by grisevg** [comment]
Is it fixed now by https://github.com/KhronosGroup/Vulkan-ValidationLayers/commit/4c7d15dda8d81df30153e0470a54e3d0ad17d2e2#diff-617ec4ceb169e2ba63e3b5ba46ef3647 ?

**Comment by mark-lunarg** [comment]
@psalz, @DiligentGraphics, can y'all check and verify that this issue is cleared up in your failure cases?  @grisevg, thank you for time and information!

**Comment by DiligentGraphics** [comment]
Is the fix going into the next release?

**Comment by mikew-lunarg** [comment]
By release, assuming you mean SDK release, if it is merged into this repo's master branch before the next SDK cycle (which hasn't started yet), then yes it will be in the next SDK.

**Comment by grisevg** [comment]
it's already released in 1.2.131

**Comment by ClemensRoegner** [comment]
Hi! 
I am running into the same issue still with 1.2.131.2.  Anybody else still hast the problem?

**Comment by psalz** [comment]
Sorry for the delay; I've only now gotten around to testing this again. I can also confirm that **as of 1.2.131.2 this has apparently not been fixed**. The test case I provided above still fails for me, and I also get the false positive in the application that initially prompted my investigation.

**Comment by Nomoresleep** [comment]
I am running into a similar issue on **1.2.131.2** as well as **1.2.135.0**. I am using two command pools (one per swapchain buffer) and using the same query pool for both of them (this does not violate the spec as far as I know but please correct me if I am wrong). If I am using only one command pool with the query pool the false positives are gone.

**Comment by TheMostDiligent** [comment]
I did not see this issue in 1.2.135.0, however I just tried the latest build from appveyor, and there is a lot of messages again.

Here is the link to the [test app](https://www.dropbox.com/s/a660qzy8edk9i60/Queries.zip?dl=0).

**Comment by TheMostDiligent** [comment]
I think this issue might be fixed in the latest build. I was not able to repro it with the latest binaries from appveyor.

**Comment by RuslanKutdusov** [comment]
We experience the same problem even on the latest build. We perform the following:
```
Reset pool#0 in cmdBuffer#0
Begin query#0 in pool#0 in cmdBuffer#0
End query#0 pool#0 in cmdBuffer#0
Submit cmdBuffer#0 with fence
Wait for fence // here we wait for the previous frame and cmdBuffer#0 to complete

Begin query#1 in pool#0 in cmdBuffer#1
End query#1 pool#0 in cmdBuffer#1
Submit cmdBuffer#1 // here we got validation layer error that query#1 not reset
```
But everything is ok if I move pool reset after 'Wait for fence' and reset pool after creation
```
vkResetQueryPool(pool#0)

Begin query#0 in pool#0 in cmdBuffer#0
End query#0 pool#0 in cmdBuffer#0
Submit cmdBuffer#0 with fence
Wait for fence // here we wait for the previous frame and cmdBuffer#0 to complete

Reset pool#0 in cmdBuffer#1
Begin query#1 in pool#0 in cmdBuffer#1
End query#1 pool#0 in cmdBuffer#1
Submit cmdBuffer#1 // here we got validation layer error that query#1 not reset
```

**Comment by xxxbxxx** [comment]
I also get the message with "vulkan-validation-layers 1.2.146" 

```
...
Wait for fence (associated with cmdBuffer#0)
vkCmdWriteTimestamp(cmdBuffer#0)
Submit cmdBuffer#0
...
Wait for fence (associated with cmdBuffer#1)
vkGetQueryPoolResults(cmdBuffer#1) -> ok
vkCmdResetQueryPool(cmdBuffer#1)
Submit cmdBuffer#1
....
Wait for fence (associated with cmdBuffer#2)
.....
and later, when using the query again --> "Queries must also be reset between uses."
```

I did try to investigate a bit and I noticed that in the validationlayer, queryToStateMap[querry] changes are:
  * `Submit cmdBuffer#0`   -> in use
  * `Submit cmdBuffer#1`   -> reset
  * `Wait for fence (associated with cmdBuffer#2)`   -> in use !

I guess that's because the query becomes available to `vkGetQueryPoolResults()` before the cmdBuffer#0 is retired during the `Wait for fence (associated with cmdBuffer#2)`


I tried this little change:
```diff
--- src/Vulkan-ValidationLayers-1.2.146/layers/state_tracker.cpp	2020-07-09 20:25:20.000000000 +0200
+++ state_tracker.cpp	2020-08-29 18:16:25.675223810 +0200
@@ -1767,6 +1767,9 @@ void ValidationStateTracker::RetireWorkO
                 if (queryStatePair.second == QUERYSTATE_ENDED) {
                     queryToStateMap[queryStatePair.first] = QUERYSTATE_AVAILABLE;
                 }
+                else if (queryStatePair.second == QUERYSTATE_RESET) {
+                    queryToStateMap[queryStatePair.first] = QUERYSTATE_RESET;
+                }
             }
             cb_node->in_use.fetch_sub(1);
         }
```

Which seems to fix the problem for me (but no idea if it's correct or if it catches all the cases)
the new behavior for queryToStateMap[querry] is:
  * `Submit cmdBuffer#0`   -> in use
  * `Submit cmdBuffer#1`   -> reset
  * `Wait for fence (associated with cmdBuffer#2)`   -> in use  (when cmdBuffer#0 is retired)
  * `Wait for fence (associated with cmdBuffer#3)`   -> reset (when cmdBuffer#1 is retired)

**Comment by mark-lunarg** [comment]
Thanks guys, will bump this in priority.

**Comment by kruseborn** [comment]
We are getting this specific validation error quite often in our Engine, is it possible to get a higher priority on this issue?

**Comment by mark-lunarg** [comment]
@ncesario-lunarg, lets get together today and come up with a plan for this one.

**Comment by ikapoura** [comment]
Hello there. The linked pull request is still open from September. Do we have any updates on it? Thank you.

**Comment by KarenGhavam-lunarG** [comment]
> Hello there. The linked pull request is still open from September. Do we have any updates on it? Thank you.

Work for the 1.3 release and the Profiles Toolset impacted completion of this PR. Now that 1.3 is released, resource will be able to get back to this.

**Comment by danginsburg** [comment]
I'm hitting this same issue in one of our games.
